Update episode03_configuration

SarahAlidoost opened the pull request:
# Pull Request checklist

We appreciate your time and effort to improve the tutorial. Please keep in mind that lesson maintainers are volunteers and it may be some time before they can respond to your contribution.

* * *

## Before you start

- [x] Read [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md).
- [x] Create an [issue](https://github.com/ESMValGroup/tutorial/issues) to discuss your idea. This allows your contributions to be incorporated into the tutorial.

## Tasks

- [x] Give this pull request a descriptive title.
- [x] If you are contributing to existing lesson materials, please make sure the content conforms to the `Lesson development` section in [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md) and does not contain any spelling or grammatical errors.
- [x] If you are making a new episode, please make sure the content conforms to the `Lesson organization` and `Lesson formatting` sections in [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md) and does not contain any spelling or grammatical errors.
- [ ] Preferably Codacy checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why.
- [x] Preview changes on your machine before pushing them to GitHub by running `make serve`, alternatively `make docker-serve`. Please see the `Previewing your changes locally` section in [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md) for installation instructions.
- [x] All code instructions have been tested.

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes  #11

ledm commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,158 @@ title: "Configuration"
-~~~
+~~~bash
+  cd ESMValTool
   cp config-user-example.yml config-user.yml
```
My workflow is to copy this to a new directory next to ESMValTool called run, or workspace. I find this works better in keeping things in one place. 

Is there a best practices document somewhere that we should consult?

SarahAlidoost replied:
@ledm Thanks. I like your suggestion. I couldn't find any best practices for this issue. However, I have checked some lessons. There is no standard way, some choose "code", "data" directories in Desktop which is not good. I applied this workflow: 

```
- esmvaltool_tutorial
   - ESMValTool (Github repo)
   - data
   - other files like a copy of config file. 
```

Instruction on how to make these directories will come in `setup` section of the tutorial.

bouweandela replied:
There is an open issue about making config-user.yml more user friendly: https://github.com/ESMValGroup/ESMValCore/issues/64

ledm commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,158 @@ title: "Configuration"
 ~~~
+
+In this lesson, we will work with data from [CMIP5](https://esgf-node.llnl.gov/projects/cmip5/).
+Add the root path of the folder where you downloaded the data during the [Setup](https://esmvalgroup.github.io/tutorial/setup.html).
```
This assumes that people will be working locally, instead of somewhere like Jasmin, where the data will be available already. 

In the long term, a local installation is useful, but which version do we want to teach in the tutorial? Local or Jasmin? @valeriupredoi @bouweandela , thoughts?

Alternatively, do we want to write this documentation twice? Once for local and once for jasmin?

bouweandela replied:
Maybe we can make the tutorial for a local installation and add commented out versions of the correct drs settings to config-user-example.yml, so people can just uncomment the lines they need if they happen to be on a cluster that is connected to an ESGF node?

SarahAlidoost replied:
@ledm @bouweandela I agree with Bouwe to make the tutorial for local use. I updated this part of the lesson. Now the instruction on how to get the data (or its rootpath) will come in `setup` section of the tutorial. This is the most common approach for several lessons that I have seen in software carpentry.  The `setup` section contains everything related to data and additional material. Therefore, the episode content won't be changed.

bouweandela replied:
Created an issue to do this: https://github.com/ESMValGroup/ESMValTool/issues/1622

bouweandela commented:
@ledm Are you OK with the changes? Can this be merged?

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
 questions:
-- "What is user configuration file and how can I use it?"
+- "What is the user configuration file and how can I use it?"
 objectives:
```
I would probably rephrase this question to "What is the user configuration file and how do I use it?" Otherwise it sounds a little too optional for me, like it would actually be possible to ignore the user configuration file and run the ESMValTool without it.

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
 objectives:
-- "Understand the data directories structure"
+- "Understand the contents of the user-config.yml file"
+- "Prepare the user-config.yml file for use"
```
I would probable rephrase this point here: "Prepare a personalized user-config.yml file" or something like it?

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
-- "Understand the data directories structure"
+- "Understand the contents of the user-config.yml file"
+- "Prepare the user-config.yml file for use"
 - "Configure ESMValTool to ignore some settings"
```
I am not sure I understand what you mean here. You want to set settings in the user-config.yml file so that the ESMValTool is ignoring some things it would normally do?

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
-- "The ``config-user.yml`` file tells ESMValTool what directory is the destination"
+- "The ``config-user.yml`` tells ESMValTool where to find input data."
+- " ``rootpath`` determines root directory for input data."
+- " ``output_dir`` is the destination directory."
```
Maybe rephrase to "``output_dir`` defines the destination directory."?

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
-- "The ``config-user.yml`` file tells ESMValTool what data are input"
-- "The ``config-user.yml`` file tells ESMValTool what directory is the destination"
+- "The ``config-user.yml`` tells ESMValTool where to find input data."
+- " ``rootpath`` determines root directory for input data."
```
Maybe rephrase to "``rootpath`` defines the root directory for the input data."

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
-Make a copy and rename it to ``config-user.yml``:
+This is an [YAML file](https://yaml.org/spec/1.2/spec.html). An example configuration file can be found in the root directory of the ESMValTool repository: [config-user-example.yml](https://github.com/ESMValGroup/ESMValTool/blob/master/config-user-example.yml).
 
+Let's download it to our working directory ``esmvaltool_tutorial`` that is made during the [Setup](https://esmvalgroup.github.io/tutorial/setup.html).
```
Maybe rephrase to "... that was created during the [Setup]..."?

bouweandela replied:
It might be good to create the link as described in the section [Locations and Names](https://carpentries.github.io/lesson-example/04-formatting/index.html), because then it will still work even if the repository is moved/renamed.

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
+Let's download it to our working directory ``esmvaltool_tutorial`` that is made during the [Setup](https://esmvalgroup.github.io/tutorial/setup.html).
+To do that, click on [this link](https://raw.githubusercontent.com/ESMValGroup/ESMValTool/master/config-user-example.yml) to see a raw version of the file, right-click and press ``save as``, then you can rename it to ``config-user.yml`` and save it into the working directory ``esmvaltool_tutorial``.
+
+Now in a terminal, let's change our working directory to ``esmvaltool_tutorial``. Then, we run a text editor called Nano to open the configuration file:
```
Maybe rephrase to "Now, let's change our working directory in a terminal window to ..."

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
 
+* Rootpath to input data
+* Directory structure for the data from different projects
+* Number of parallel tasks
```
Maybe rephrase to "* Number of tasks that can be run in parallel"

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
+* Auxiliary data directory
+* Output settings
+
+> ## Which text editor
```
I am not sure why you called this section "Which text editor". With this header I would have expected some text about the pros and cons of different text editors. Maybe rephrase the header to "Text editor side note" or something

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
+
+> ## Which text editor
+>
+> No matter what editor you use, you will need to know where it searches for and saves files. If you start it from the shell, it will (probably) use your current working directory as its default location. We use ``nano`` in examples because it is one of the least complex text editors. Press <kbd>ctrl</kbd> + <kbd>O</kbd> to save the file, and then <kbd>ctrl</kbd> + <kbd>X</kbd> to exit ``nano``.
```
I would probably rephrase to "... We use ``nano`` in examples here because...".

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
+> ## Which text editor
+>
+> No matter what editor you use, you will need to know where it searches for and saves files. If you start it from the shell, it will (probably) use your current working directory as its default location. We use ``nano`` in examples because it is one of the least complex text editors. Press <kbd>ctrl</kbd> + <kbd>O</kbd> to save the file, and then <kbd>ctrl</kbd> + <kbd>X</kbd> to exit ``nano``.
+{: .callout}
```
This command is still visible in the text in it's formatted form. Shouldn't be, right?

SarahAlidoost replied:
It is shown in the correct style on the webpage. Could you please check it via [previewing your changes locally](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md#previewing-your-changes-locally)

hb326 replied:
Sorry, I have trouble today with everything, as it seems. Can't get it to work. Maybe @LisaBock can have a look if it is displayed correctly?

LisaBock replied:
Yes, looks correct.

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
+## Rootpath to input data
+
+ESMValTool uses several categories (in ESMValTool, this is referred to as projects) for input data based on their source.
+The current categories in the configuration file are mentioned below. For example, CMIP is used for a dataset from the climate model intercomparison project whereas OBS for an observational dataset. The ``rootpath`` specifies the directories where ESMValTool will look for input data. For each category, you can define either one path or several paths as a list.
```
I would change the wording to "... CMIP is used for a dataset from the Climate Model Intercomparison Project whereas OBS is used for an observational dataset."

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
+  ana4mips: ~/ana4mips_inputpath
+  native6:  ~/native6_inputpath
+  RAWOBS: ~/rawobs_inputpath
+  default: ~/default_inputpath
```
Do we want to add an explanation about the other projects mentioned here beside CMIP and OBS? Or do we want to refer to a different episode where these projects might be described (like how to add your own observational dataset)?

SarahAlidoost replied:
good questions. some points for a discussion. I need to check with other episodes.

SarahAlidoost replied:
I added a link to the esmvaltool documentation about other projects.

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
 ~~~
+
+In this lesson, we will work with data from [CMIP5](https://esgf-node.llnl.gov/projects/cmip5/).
+We add the root path of the folder where data is available.
```
Maybe rephrase to "We add the root path of the folder where our/your data is available." I am not sure which one of the pronouns I would prefer, but I think it would clarify that the path to the student's data should be added here.

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
+> ## Setting the correct rootpath
+>
+> * To get the data (or its correct rootpath), check instruction in [Setup](https://esmvalgroup.github.io/tutorial/setup.html).
+> * For more information about setting the rootpath, you can visit ESMValTool [documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/datafinder.html).
```
I would rephrase to "For more informatino about setting the roothpath, see also ESMValTool...".

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
-  auxiliary_data_dir: ~/auxiliary_data
+> ## Setting the correct rootpath
+>
+> * To get the data (or its correct rootpath), check instruction in [Setup](https://esmvalgroup.github.io/tutorial/setup.html).
```
I am not sure what you mean here. Is this a pointer toward instructions on where to actually get the CMIP5 data or is this a pointer toward the correct setup of the rootpath?

SarahAlidoost replied:
The instruction on how to get the data (or its rootpath) will come in setup section of the tutorial. This is the most common approach for several lessons that I have seen in software carpentry. The setup section contains everything related to data and additional material. Therefore, the episode content won't be changed. Please also see the discussion in the comment [here](https://github.com/ESMValGroup/tutorial/pull/12#discussion_r393092548)

hb326 replied:
Ok.

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
+>
+> * To get the data (or its correct rootpath), check instruction in [Setup](https://esmvalgroup.github.io/tutorial/setup.html).
+> * For more information about setting the rootpath, you can visit ESMValTool [documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/datafinder.html).
+{: .callout}
```
This command is also visible in the formatted text.

SarahAlidoost replied:
It is shown in the correct style on the webpage. Could you please check it via [previewing your changes locally](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md#previewing-your-changes-locally)

LisaBock replied:
Format looks correct, but I would remove the three dots after 'ESMValTool' and before '[documentation]'.

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
+## Directory structure for the data from different projects
 
+Input data can be from various models, observations and reanalysis data that adhere to the [CF/CMOR standard](https://cmor.llnl.gov/). To set a directory, you can use one of the values of ``default``, ``BADC``, ``DKRZ``, ``ETHZ``, .... Let's use ``default`` in our example:
 
```
I think we might have to mention here that the information given here for "drs" refers to a folder structure that describes how the data is stored. Or would that be too much information for beginners here? If we don't mention it, I think it is not clear to the students what the difference between "BADC" and "ETHZ" would mean.

SarahAlidoost replied:
I added more explanations.

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
+## Number of parallel tasks
 
-.. code-block:: yaml
+This option enables you to perform parallel processing. You can choose the number of tasks in parallel as 1/2/3/4/... or you can set it to ``null`` that tells ESMValTool to use the number of available CPUs:
```
I would rephrase to "... or you can set it to ``null``. That tells ESMValTool...".

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
+## Number of parallel tasks
 
-.. code-block:: yaml
+This option enables you to perform parallel processing. You can choose the number of tasks in parallel as 1/2/3/4/... or you can set it to ``null`` that tells ESMValTool to use the number of available CPUs:
```
Does it tell the ESMValTool to use the maximum number of available GPUs if it is set to ``null``?

SarahAlidoost replied:
I revised the text here. Please also see the e[smvaltool documentation ](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/quickstart/configure.html?highlight=parallel#user-configuration-file) where mentions that "_Set to null to use the number of available CPUs."_

hb326 replied:
OK.

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
-output, i.e. figures, data, logs, etc. Make sure to set appropriate paths.
+> ## Set the number of tasks
+>
+> If you run out of memory, try setting ``max_parallel_tasks`` to 1. Then, check the amount of memory you need for that by inspecting the file ``run/resource_usage.txt`` in the output directory. Using the number there you can increase the number of parallel tasks again to a reasonable number for the amount of memory available in your system.
```
This is probably also something that we should link to the "debugging" episode.

SarahAlidoost replied:
@hb326 thank you. I will consider it for the debugging episode.

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
-.. code-block:: yaml
+## Destination directory
 
+The destination directory is the rootpath where ESMValTool will store its output, i.e. figures, data, logs, etc. With every run, ESMValTool automatically generates a new output folder determined by the recipe name, and date and time using the format: YYYYMMDD_HHMMSS. This folder contains four further subfolders: ``plots``, ``preproc``, ``run``, ``work``.
```
I would rephrase to "... by recipe name, date and time ..."

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
 
+The destination directory is the rootpath where ESMValTool will store its output, i.e. figures, data, logs, etc. With every run, ESMValTool automatically generates a new output folder determined by the recipe name, and date and time using the format: YYYYMMDD_HHMMSS. This folder contains four further subfolders: ``plots``, ``preproc``, ``run``, ``work``.
 
+Let's name our destination directory as ``esmvaltool_output`` in the working directory:
```
I would rephrase to "... our destinateion directory ``esmvaltool_output`` in ..."

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
+> * ``plots``: the location for all plots, split by individual diagnostics and fields.
+> * ``preproc``: this folder contains all the preprocessed data and metadata.yml interface files. Note that by default this directory will be deleted after each run because most users will only need the results from the diagnostic scripts.
+> * ``run``: this folder includes all log files, a copy of the recipe, a summary of the resource usage, and the settings.yml interface files, resource_usage.txt and temporary files created by the diagnostic scripts.
+> * ``work``: a place for any diagnostic script results that are not plots, e.g. files in NetCDF format (depends on the diagnostic script).
```
I would rephrase to be consistent with the entries before "``work``: this folder is a place for any..."

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
-   The ``config-user.yml`` file is specified as argument at run time, so it is
-   possible to have several available with different purposes: one for
-   formalised runs, one for debugging, etc...
+The ``auxiliary_data_dir`` setting is the path to place any required additional auxiliary data files. This location allows us to tell the diagnostic script where to find the files if they can not be downloaded at runtime. This option is not for model or observational datasets, rather it is for data files used in plotting such as coastline descriptions and so on.
```
I would rephrase "... is the path where any additionally required auxiliary data files are stored." And "This option should not be used for model or observational datasets, but for data files (e.g. shape files) used in plotting, such as ...".

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
-{% include links.md %}
+## Output settings
 
+These settings are used to inform ESMValTool about your preference. You can turn on or off the setting by ``true`` or ``false`` values. Most of these settings are fairly self-explanatory, ie:
```
Maybe rephrase to "These settings are used to inform ESMValTool about your preference about specific actions. "

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
+> It is possible to have several configuration files with different purposes, for example: config-user_formalised_runs.yml, config-user_debugging.yml
+{: .callout}
+
+> ## Different settings
```
Is this already one of the tasks for people to try out? It is not clear to me from the placing...

SarahAlidoost replied:
I revised the title. This section will be shown as a challenge. A learner needs to answer the question. The solution is also provided.

hb326 replied:
Ok.

hb326 commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,178 @@ title: "Configuration"
+
+> ## Different settings
+>
+> In the configuration file, which settings are useful to store preprocessed data?
```
I would probably rephrase to "In the configuration file, which settings are usefule to make sure preprocessed data is stored when ESMValTool is run?

sverhoeven commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,236 @@ title: "Configuration"
+[config-user-example.yml](https://github.com/ESMValGroup/ESMValTool/blob/master/config-user-example.yml).
+
+Let's download it to our working directory ``esmvaltool_tutorial``
+that was created during the [Setup]({{ page.root }}{% link setup.md %}).
```
The setup.md is of #46 and the installation instructions of #31 , don't create an `esmvaltool_tutorial` directory. Can the directory reason and creation be described here or in  #31 or in #46?

SarahAlidoost replied:
now the directory creation is explained inside this lesson.

sverhoeven commented on _episodes/03-configuration.md:
```diff
@@ -3,91 +3,236 @@ title: "Configuration"
+``esmvaltool_tutorial``.
+
+Now, let's change our working directory in a terminal window to ``esmvaltool_tutorial``.
+Then, we run a text editor called Nano to open the configuration file:
```
Why am I opening the config file in an editor?
```suggestion
Then, we run a text editor called Nano to have a look inside the configuration file:
```

SarahAlidoost commented:
@sverhoeven, @LisaBock, @hb326, @bouweandela thank you for your helpful comments. I revised the text addressing your comments. It seems that some codacy issues cannot be fixed at this moment. Can we merge this pull request?

hb326 commented:
> 
> 
> @sverhoeven, @LisaBock, @hb326, @bouweandela thank you for your helpful comments. I revised the text addressing your comments. It seems that some codacy issues cannot be fixed at this moment. Can we merge this pull request?

Fine with me!
